client/src/commands/getTcb.ts

```diff
@@ -0,0 +1,104 @@
+    this.lines = source.split('\n').map(line => {
+      const line_len = line.length + 1 /* ='\n' */;
+      const offset_and_len: [number, number] = [offset, line_len];
+      offset += line_len;
+      return offset_and_len;
+    });
+  }
+
+  toPosition(offset: number): vscode.Position {
+    const lineno = this.lines.findIndex(
+        ([line_offset, line_len]) => offset >= line_offset && offset < line_offset + line_len);
+    if (lineno === -1) throw new Error(`Bad state: ${offset} is out of bounds`);
+    const [line_offset, ] = this.lines[lineno];
+    return new vscode.Position(lineno, offset - line_offset);
+  }
+}
+
+async function executeGetTemplateTcb(
+    ngClient: AngularLanguageClient, didRequestTcb: vscode.EventEmitter<vscode.Uri>) {
+  const activeEditor = vscode.window.activeTextEditor;
+  if (!activeEditor) {
+    return;
+  }
+  const client = ngClient.languageClient;
+  if (client === null) {
+    return;
+  }
+
+  const tcb = await client.sendRequest(lspext.GetTcb, {
+    textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(activeEditor.document),
+    position: activeEditor.selection.active,
+  });
+  if (tcb === undefined) {
+    return;
+  }
+
+  // Allocate a provider of documents corresponding to the `ng` URI scheme,
+  // which we will use to provide a virtual document with the TCB contents.
+  const tcbProvider: vscode.TextDocumentContentProvider = {
+    onDidChange: didRequestTcb.event,
+    provideTextDocumentContent: () => tcb.content,
+  };
+  const disposeNgProvider = vscode.workspace.registerTextDocumentContentProvider('ng', tcbProvider);
+  const tcbUri = vscode.Uri.parse('').with({
+    scheme: 'ng',
+    authority: 'getTcb',
+    path: tcb.fileName,
+  });
+
+  const document = await vscode.workspace.openTextDocument(tcbUri);
+  // Notify VSCode of a change to the TCB virtual document, prompting it to
+  // re-evaluate the document content. This is needed to bust VSCode's document
+  // cache if someone requests a TCB that we previously opened.
+  // https://code.visualstudio.com/api/extension-guides/virtual-documents#update-virtual-documents
+  didRequestTcb.fire(tcbUri);
+  const editor = await vscode.window.showTextDocument(document, {
+    viewColumn: (vscode.window.activeTextEditor?.viewColumn ?? 1) + 1,
+    preserveFocus: true,
+  });
+
+  const sm = new SourceMap(tcb.content);
+  const highlightSelections = tcb.selections.map(
+      ({start, length}) => new vscode.Range(sm.toPosition(start), sm.toPosition(start + length)));
+  editor.setDecorations(TCB_HIGHLIGHT_DECORATION, highlightSelections);
+
+  // We immediately dispose of the provider so we can create a fresh provider
+  // for the `ng` scheme next time `getTcb` is requested.
+  disposeNgProvider.dispose();
+}
+
+/**
+ * Command getTemplateTcb displays a typecheck block for the template a user has
+ * an active selection over, if any.
+ */
+export function getTemplateTcb(ngClient: AngularLanguageClient) {
```

Comment:
I could find no way to test the client-side behavior on the VSCode docs. Let me know if you have any ideas; for now I recommend just spinning up a session and see what happens. On my side it looked to be okay, I can add a demo if folks desire.

Reply:
yeah, we don't have client-side tests / can't run them on CI.
I think a manual test here is sufficient, since this feature is not user-facing.